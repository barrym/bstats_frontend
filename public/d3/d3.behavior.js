(function(){d3.behavior = {};
// TODO unbind zoom behavior?
// TODO unbind listener?
d3.behavior.zoom = function() {
  var x = 0,
      y = 0,
      z = 0,
      panning,
      zooming,
      locations = {}, // identifier -> location
      last = 0,
      target,
      targuments,
      event = d3.dispatch("zoom");

  // mousewheel events are totally broken!
  // https://bugs.webkit.org/show_bug.cgi?id=40441
  // not only that, but Chrome and Safari differ in re. to acceleration!
  var div = d3.select("body").append("div")
      .style("visibility", "hidden")
      .style("top", 0)
      .style("height", 0)
      .style("width", 0)
      .style("overflow-y", "scroll")
    .append("div")
      .style("height", "2000px")
    .node().parentNode;

  function zoom() {
    this
        .on("mousedown.zoom", mousedown)
        .on("mousewheel.zoom", mousewheel)
        .on("DOMMouseScroll.zoom", dblclick)
        .on("dblclick.zoom", dblclick)
        .on("touchstart.zoom", touchstart);

    d3.select(window)
        .on("mousemove.zoom", mousemove)
        .on("mouseup.zoom", mouseup)
        .on("touchmove.zoom", touchmove)
        .on("touchend.zoom", touchup);
  }

  function location(point) {
    return [point[0] - x, point[1] - y, z];
  }

  // snapshot the target, data and index for subsequent dispatch
  function start() {
    target = this;
    targuments = arguments;
  }

  // Note: Since we don't rotate, it's possible for the touches to become
  // slightly detached from their original positions. Thus, we recompute the
  // touch points on touchend as well as touchstart!
  function touchup() {
    var touches = d3.svg.touches(target),
        i = -1,
        n = touches.length,
        touch;
    while (++i < n) locations[(touch = touches[i]).identifier] = location(touch);
    return touches;
  }

  function touchstart() {
    start.apply(this, arguments);
    var touches = touchup(),
        touch,
        now = Date.now();

    // doubletap detection
    if ((touches.length === 1) && (now - last < 300)) {
      var touch = touches[0];
      zoomto(1 + Math.floor(z), touch, locations[touch.identifier]);
    }
    last = now;
  }

  function touchmove() {
    var touches = d3.svg.touches(target);
    switch (touches.length) {

      // single-touch pan
      case 1: {
        var touch = touches[0];
        zoomto(z, touch, locations[touch.identifier]);
        break;
      }

      // double-touch pan + zoom
      case 2: {
        var p0 = touches[0],
            p1 = touches[1],
            p2 = [(p0[0] + p1[0]) / 2, (p0[1] + p1[1]) / 2],
            l0 = locations[p0.identifier],
            l1 = locations[p1.identifier],
            l2 = [(l0[0] + l1[0]) / 2, (l0[1] + l1[1]) / 2, l0[2]];
        zoomto(Math.log(d3.event.scale) / Math.LN2 + l0[2], p2, l2);
        break;
      }
    }
  }

  function mousedown() {
    start.apply(this, arguments);
    panning = location(d3.svg.mouse(target));
    d3.event.preventDefault();
    window.focus();
  }

  function mousemove() {
    zooming = null;
    if (panning) zoomto(z, d3.svg.mouse(target), panning);
  }

  function mouseup() {
    if (panning) {
      mousemove();
      panning = null;
    }
  }

  function mousewheel() {
    start.apply(this, arguments);

    // store starting mouse location
    if (!zooming) zooming = location(d3.svg.mouse(target));

    // detect the pixels that would be scrolled by this wheel event
    var e = d3.event, delta;
    try {
      div.scrollTop = 1000;
      div.dispatchEvent(e);
      delta = 1000 - div.scrollTop;
    } catch (error) {
      delta = e.wheelDelta || -e.detail;
    }

    zoomto(delta * .005 + z, d3.svg.mouse(target), zooming);
  }

  function dblclick() {
    start.apply(this, arguments);
    var mouse = d3.svg.mouse(target);
    zoomto(d3.event.shiftKey ? Math.ceil(z - 1) : Math.floor(z + 1), mouse, location(mouse));
  }

  function zoomto(z1, x0, x1) {
    var k = Math.pow(2, (z = z1) - x1[2]);
    x = x0[0] - k * x1[0];
    y = x0[1] - k * x1[1];
    dispatch();
  }

  function dispatch() {
    var o = d3.event, // Events can be reentrant (e.g., focus).
        k = Math.pow(2, z);

    d3.event = {
      scale: k,
      translate: [x, y],
      transform: function(sx, sy) {
        if (sx) transform(sx, x);
        if (sy) transform(sy, y);
      }
    };

    function transform(scale, o) {
      var domain = scale.__domain || (scale.__domain = scale.domain()),
          range = scale.range().map(function(v) { return (v - o) / k; });
      scale.domain(domain).domain(range.map(scale.invert));
    }

    try {
      event.zoom.dispatch.apply(target, targuments);
    } finally {
      d3.event = o;
    }

    o.preventDefault();
  }

  zoom.on = function(type, listener) {
    event[type].add(listener);
    return zoom;
  };

  return zoom;
};
})();
