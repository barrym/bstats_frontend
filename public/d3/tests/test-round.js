require("./../lib/env-js/envjs/node");
require("./../d3");

console.log("degenerate:");
console.log("          0 ->", d3.round(0));
console.log("        NaN ->", d3.round(NaN));
console.log("   Infinity ->", d3.round(Infinity));
console.log("  -Infinity ->", d3.round(-Infinity));
console.log("");

console.log("default:");
console.log("      10.51 ->", d3.round(10.51));
console.log("      10.50 ->", d3.round(10.50));
console.log("      10.49 ->", d3.round(10.49));
console.log("       0.51 ->", d3.round(0.51));
console.log("       0.50 ->", d3.round(0.50));
console.log("       0.49 ->", d3.round(0.49));
console.log("      -0.51 ->", d3.round(-0.51));
console.log("      -0.50 ->", d3.round(-0.50));
console.log("      -0.49 ->", d3.round(-0.49));
console.log("     -10.51 ->", d3.round(-10.51));
console.log("     -10.50 ->", d3.round(-10.50));
console.log("     -10.49 ->", d3.round(-10.49));
console.log("");

console.log("precision 1:");
console.log("      10.51 ->", d3.round(10.51, 1));
console.log("      10.50 ->", d3.round(10.50, 1));
console.log("      10.49 ->", d3.round(10.49, 1));
console.log("       0.51 ->", d3.round(0.51, 1));
console.log("       0.50 ->", d3.round(0.50, 1));
console.log("       0.49 ->", d3.round(0.49, 1));
console.log("      -0.51 ->", d3.round(-0.51, 1));
console.log("      -0.50 ->", d3.round(-0.50, 1));
console.log("      -0.49 ->", d3.round(-0.49, 1));
console.log("     -10.51 ->", d3.round(-10.51, 1));
console.log("     -10.50 ->", d3.round(-10.50, 1));
console.log("     -10.49 ->", d3.round(-10.49, 1));
console.log("");

console.log("precision 2:");
console.log("      10.51 ->", d3.round(10.51, 2));
console.log("      10.50 ->", d3.round(10.50, 2));
console.log("      10.49 ->", d3.round(10.49, 2));
console.log("       0.51 ->", d3.round(0.51, 2));
console.log("       0.50 ->", d3.round(0.50, 2));
console.log("       0.49 ->", d3.round(0.49, 2));
console.log("      -0.51 ->", d3.round(-0.51, 2));
console.log("      -0.50 ->", d3.round(-0.50, 2));
console.log("      -0.49 ->", d3.round(-0.49, 2));
console.log("     -10.51 ->", d3.round(-10.51, 2));
console.log("     -10.50 ->", d3.round(-10.50, 2));
console.log("     -10.49 ->", d3.round(-10.49, 2));
console.log("");

console.log("precision -1:");
console.log("     123.45 ->", d3.round(123.45, -1));
console.log("     345.67 ->", d3.round(345.67, -1));
console.log("    -123.45 ->", d3.round(-123.45, -1));
console.log("    -345.67 ->", d3.round(-345.67, -1));
console.log("");

console.log("precision -2:");
console.log("     123.45 ->", d3.round(123.45, -2));
console.log("     456.78 ->", d3.round(456.78, -2));
console.log("    -123.45 ->", d3.round(-123.45, -2));
console.log("    -456.78 ->", d3.round(-456.78, -2));
console.log("");

console.log("precision -3:");
console.log("     123.45 ->", d3.round(123.45, -3));
console.log("     567.89 ->", d3.round(567.89, -3));
console.log("    -123.45 ->", d3.round(-123.45, -3));
console.log("    -567.89 ->", d3.round(-567.89, -3));
console.log("");
