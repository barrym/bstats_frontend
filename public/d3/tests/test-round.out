degenerate:
          0 -> 0
        NaN -> NaN
   Infinity -> Infinity
  -Infinity -> -Infinity

default:
      10.51 -> 11
      10.50 -> 11
      10.49 -> 10
       0.51 -> 1
       0.50 -> 1
       0.49 -> 0
      -0.51 -> -1
      -0.50 -> 0
      -0.49 -> 0
     -10.51 -> -11
     -10.50 -> -10
     -10.49 -> -10

precision 1:
      10.51 -> 10.5
      10.50 -> 10.5
      10.49 -> 10.5
       0.51 -> 0.5
       0.50 -> 0.5
       0.49 -> 0.5
      -0.51 -> -0.5
      -0.50 -> -0.5
      -0.49 -> -0.5
     -10.51 -> -10.5
     -10.50 -> -10.5
     -10.49 -> -10.5

precision 2:
      10.51 -> 10.51
      10.50 -> 10.5
      10.49 -> 10.49
       0.51 -> 0.51
       0.50 -> 0.5
       0.49 -> 0.49
      -0.51 -> -0.51
      -0.50 -> -0.5
      -0.49 -> -0.49
     -10.51 -> -10.51
     -10.50 -> -10.5
     -10.49 -> -10.49

precision -1:
     123.45 -> 120
     345.67 -> 350
    -123.45 -> -120
    -345.67 -> -350

precision -2:
     123.45 -> 100
     456.78 -> 500
    -123.45 -> -100
    -456.78 -> -500

precision -3:
     123.45 -> 0
     567.89 -> 1000
    -123.45 -> 0
    -567.89 -> -1000

